### ISSUE TITLE
Enhancement: event triggered Network

### ISSUE THREAD

**Issue by wongck68** [freemint/freemint#53]
Since most of the GEM programs are using event_multi ( or other event functions), would it be nice to have an network data ready event. 
Normally a GEM program seems hanging while waiting for network data. If this were an event, it will still be able to service other events and make the program more responsive. 

rgds
wongck

**Comment by mikrosk** [comment]
Usually such stuff is handled in a separate thread and yes, the main thread is then notified. I hope to look at the thread situation in FreeMiNT some day, to this moment I have no idea whether/how much are threads supported and what is needed. Very blurry topic every time someone brings it up.

**Comment by mfro0** [comment]
You can do the same with Pfork() already (although with a little overhead due to the duplicated data and bss segment). 

Transport the ap_id of your main program to the forked networking process by some means (e.g. file or shared memory) and just do an `appl_init()` call there.
You can then send a message to your main program (with `appl_write()`) whenever a networking event happens in your networking process.

**Comment by ps68060** [comment]
I have noticed that if I have no network cable plugged in, during the boot process, dhcp will just wait forever.

**Comment by atic-atac** [comment]
You should background the dhcp process using the & symbol

**Comment by wongck68** [comment]
Will Pfork() duplicate the entire parent process?
If application parent size is 500 KB as seen in task scheduler, I wonder how many child instances will it take before the system slows down,

**Comment by wongck68** [comment]
On my Atari, if there is no dhcp offers --  I get an NO DHCPOFFER message and it tries to load from previous saved IP addr, if there are none then the the return will be 0.0.0.0 address on the adapter.
All these takes around 4-5 seconds and booting continues to the desktop.
I do not need to & it.

**Comment by mfro0** [comment]
Are you sure? 

At least the comments in `fork_region()` appear to imply otherwise: https://github.com/freemint/freemint/blob/master/sys/memory.c

**Comment by th-otto** [comment]
Hm, you are right, i only saw that the loop goes over all regions. But i think this actually not quite right. Imagine code like this:
``
        .text
        .global main,_main
main:
_main:
func:
        nop
offset: .dc.l var-func
        nop
        .data
var: .dc.l 100
``

The value at offset will be "8" in this case. But this is of course only true when text+data are continguous, if only the data segment is copied upon fork, and afterwards is at a totally different location, the assumption the linker made when calculating that value isn't true anymore.

**Comment by mfro0** [comment]
I'm flabbergasted: It appears each spawned process gets its .bss and .data segment copied to where it expects it on each context switch. Frankly, I didn't know (and didn't expect) that:

```
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <mintbind.h>

int g;

int main(int argc, char *argv[])
{
    int num_children = atoi(argv[1]);
    int i;
    long p;

    for (i = 0; i < num_children; i++)
    {
        p = Pfork();
        if (p == 0L)
        {
            /* child */
            int dir = Pgetpid() & 1;
            if (dir)
            {
                for (g = 0; g < 100; g++)
                {
                    printf("&g=%p, g=%d main=%p\r\n", &g, g, main);
                    sleep(1);
                }
            }
            else
            {
                for (g = 99; g >= 0; g--)
                {
                    printf("&g=%p, g=%d main=%p\r\n", &g, g, main);
                    sleep(1);
                }
            }
            Pterm(0);
        }
        else if (p > 0)
        {
            /* parent */
            printf("child %ld successfully spawned\r\n", p);
        }
        else
            perror("Pfork:");
    }
    return 0;
}                    
```

**Comment by vinriviere** [comment]
> I'm flabbergasted: It appears each spawned process gets its private copy of its .bss and .data segment copied to where it expects it (into the parent process' address space) on each context switch.

Exactly. I saw that when patching for ColdFire. It is a shame that the key of UNIX API is based on fork(), which can't be naively implemented on non-virtual-memory systems (or processors). FreeMiNT does quite a miracle here, but I suspect that the overhead of forked process is very high (should be benchmarked). Fortunately, forked processes usually call exec() very soon, so the the situation of copy-on-context-switch should stop very quickly (would be worth to verify).

**Comment by th-otto** [comment]
> I'm flabbergasted: It appears each spawned process gets its private copy of its .bss and .data segment copied to where it expects it

Wow. That's of course a lot of overhead. Didn't know that, either. At least that explains why no-one tried to implement threads so far ;)

**Comment by vinriviere** [comment]
Of course efficient, simple threads could be implemented, without support for virtual memory or fork() tricks. But I don't know if they could match any existing well-known API.

**Comment by th-otto** [comment]
For creating a thread you might not need any fork tricks. But what happens if you do a fork() from within a thread? Or call slbopen() to open a shared lib?

**Comment by vinriviere** [comment]
This thread discussion could be moved to another issue "Add thread support to FreeMiNT".

**Comment by mfro0** [comment]
> Of course efficient, simple threads could be implemented, without support 
> for virtual memory or fork() tricks. 

Of course you could do a (relativly) simple and (relatively) "clean" threads implementation using MiNT system functions only. Using the `Tmalarm()` system call, for example, which would call a signal handler every so many milliseconds that would do a simple "thread switch" using `setjmp()`/`longjmp()` pairs combined with a `Psigreturn()` call.

But I would doubt that would help much.

IMHO, the core of @wongck68 's issue is that he wants something that alerts him when there is something to do on a file descriptor (socket). Usually, this is done using the `Fselect()` system call that can either poll (with a timeout of 0) or block (which would obviously also block the AES event loop) the process when there is nothing available. Polling is something @wongck68 wants to avoid, as far as I understand.

If we'd implement the simple threading mechanism described above, it should be possible to have one "thread" sleep on an `Fselect()` call while the other "main thread" does its AES event loop stuff until signalled somehow (by an `appl_write()`, for example) with newly received data.

That could probably work, but I would still consider it polling. You'd basically not poll "directly" using `Fselect()`anymore, but - with our threading mechanism - poll if `Fselect()` returned (and you wouldn't even have a chance to really detect if it hasn't - another timeslice wasted).

A lot of effort to basically gain nothing.

My understanding is, what @wongck68 really needs is a signal that triggers a signal handler when something is available on a file descriptor. Something like that appears to be available in MiNT (the `SIGIO` signal), but as far as I understand, the only driver implementing it currently is the ppp driver (doesn't help much as well).

I have come to the opinion that this issue should be probably rephrased to "implement SIGIO for sockets"?

**Comment by mikrosk** [comment]
By the way, a long time ago @pmandin wrote [a nice document](https://web.archive.org/web/20050122122913/http://membres.lycos.fr/pmandin/index.php?page=perso-shlibs&lang=en) about shared code/libraries on Atari.

**Comment by th-otto** [comment]
Yes, i know that document, and i totally agree. The new executable format would most likely be elf. Only problem: no-one ever tried to implement that for MiNT.

**Comment by wongck68** [comment]
SIGIO would be a nice addition.

But how does signal works in conjunction with AES events ( i.e. it's like having 2 events system now) ?
I know when a signal is received, the signal handler will be invoked.... but what happens if this is in the middle of doing a long series of event like user dragging a window (window redraw event being send as it is dragged by AES).  Will the program return to the correct point when the signal handler returns i.e. half way servicing the window dragging ?

My guess is that if it is an AES event, then the AES event only triggers at event_multi() call.... so not in the middle of servicing an event.

May be I do not understand at which part of the program does the signal is being caught.

**Comment by mfro0** [comment]
> I know when a signal is received, the signal handler will be invoked.... but what happens if this is in the middle of doing a long series of event like user dragging a window (window redraw event being send as it is dragged by AES). Will the program return to the correct point when the signal handler returns i.e. half way servicing the window dragging ?

The signal handler would handle the signal (do whatever networking stuff is needed, read or write data to the socket/internal buffers) and - when finished - signal the GUI (using `appl_write()`, for example) that something has changed and needs to be displayed.
You might also be able to hand over the handling of the network event directly to the AES event loop (using `appl_write()` without doing any networking stuff) and handle the network actions there. But I wouldn't consider that proper design (I see the AES event loop for UI handling only).

Besides implementing SIGIO, we'd also need to clarify if it's o.k. to use `appl_write()` from a signal handler. Normally, it's not (you are not allowed to do AES calls from there), but I would assume (although never tried myself) it should be safe with some caution in this exceptional case.

**Comment by mfro0** [comment]
``using non-blocking networking within AES timer (i think every 300 ms) 

this might work for small and simple programs, but it's far from optimal. You'll never achieve full bandwidth since your timer event is either triggering too often (producing "waste of heat") or not often enough (with network lags or even timeouts). You might be able to optimise that a little (with "adaptive" timer values), but for more complex applications, I suppose this is a dead end.

**Comment by mfro0** [comment]
Here is some test code that shows that above approach is indeed feasible. Because of the lack of a `SIGIO` signal, it uses `Talarm()` to send a `SIG_ALRM` signal instead that causes the signal handler to be called every second. In turn, the signal handler rearms itself and sends a message to the AES event loop (note that the docs say you are not supposed to do AES calls from a signal handler, nevertheless it appears to work fine in this case).

```
#include <stdio.h>
#include <stdlib.h>
#include <mintbind.h>
#include <stdbool.h>
#include <gem.h>

#define SIG_ALRM    14

#define ALERT       4711

static short ap_id;

void alert_handler(long signal)
{
    static short counter = 0;
    short msg[] = { ALERT, 0 };
    short msg_len;
    
    msg[1] = counter++;
    msg_len = sizeof(msg) / sizeof(short);
    appl_write(ap_id, msg_len, msg);
    
    Talarm(1);
    Psignal(SIG_ALRM, alert_handler);
}

int main(int argc, char *argv[])
{
    short butdown;
    short msgbuff[10];
    long msec;
    short mx, my, ret;
    short keystate, keyret;
    short event;
    bool quit = false;

    ap_id = appl_init();

    Talarm(1);
    Psignal(SIG_ALRM, alert_handler);

    do {
        event = evnt_multi(MU_MESAG,
                           1, 1, butdown,
                           0, 0, 0, 0, 0,
                           0, 0, 0, 0, 0,
                           msgbuff,
                           msec,
                           &mx, &my, &ret,
                           &keystate, &keyret, &ret);
        if (event & MU_MESAG)
        {
            switch (msgbuff[0])
            {
                case ALERT:
                    printf("alert signal (%d) received from signal handler\n", msgbuff[1]);
                    if (msgbuff[1] == 99)
                        quit = true;
                    break;
                default:
                    printf("unknown message %d received\n", msgbuff[0]);
                    break;
            }
        };

    } while (!quit);

    appl_exit();

    return 0;
}
```

**Comment by mikrosk** [comment]
@mfro0 just to clarify (I'm not really at home when it comes to this signal magic): this would execute `alert_handler` every second, right. So we're looping in the event loop, having fun in the UI and every second we end up in `alert_handler`. But if this handler is blocking (lasting for several seconds), the UI would be still frozen, right? Or this code was only to demonstrate the signalling and in real code I'd run `alert_handler` in a separate, forked, process?

Btw guys did you know MagiC has a "real" thread API: http://toshyp.atari.org/en/00b012.html ? (I bumped into this when I realised that this signal stuff is usable only in MiNT).

**Comment by mikrosk** [comment]
Further information about MagiC signals: _First of all it must be realised that in MagiC, AES and VDI calls from within a signal-handler are permitted (quite in contrast to MiNT)._ (http://toshyp.atari.org/en/005007.html#Signal-handler_20in_20MagiC) ... so in MagiC it is not only permitted but according to @mfro0 it also works in FreeMiNT. So now only to think of a `Talarm` replacement in MagiC and the code above should be usable in both OSes.

**Comment by mfro0** [comment]
> So we're looping in the event loop, having fun in the UI and every second we end up in alert_handler. But if this handler is blocking (lasting for several seconds), the UI would be still frozen, right?

**Comment by th-otto** [comment]
>Btw guys did you know MagiC has a "real" thread API

Yes, of  course. One of the reasons why all those mt_xxx functions were introduced in gemlib. Still a bit useless though, since i've never seen any real application making use of it, except for some demo programs. But it's not surprising, since there is no thread-safe standard library, so you would have to protect every critical operation yourself.

**Comment by mfro0** [comment]
> First of all it must be realised that in MagiC, AES and VDI calls from within a signal-handler are permitted (quite in contrast to MiNT).

I would _assume_ that many (if not most) of this information is outdated and merely applies to MiNT + MultiTOS. My understanding is that XaAES (since its a kernel process) should be able to avoid the said shortcomings.

Regarding threads: I'm in doubt they would help us much (as pointed out above already). A fully-functional thread library would be required to wrap all system calls that could potentially block into their nonblocking counterparts, which would require multi-threaded programs to ask all its file descriptors if data is available most of the time. Every system call would basically loop busy waiting until there is something to do.

Not very effective with hardware that has very limited processing power already.

**Comment by wongck68** [comment]
>  this would execute alert_handler every second, right. So we're looping in the event loop, having fun in the UI and every second we end up in alert_handler. But if this handler is blocking (lasting for several seconds), the UI would be still frozen, right? Or this code was only to demonstrate the signalling and in real code I'd run alert_handler in a separate, forked, process?

In any case, this will happen when a big lump of data is ready for reading from the network reqardless. So only  improvement is when we are waiting for the data to be available ( or during network lag) there is no hanging feeling.

**Comment by mfro0** [comment]
> In any case, this will happen when a big lump of data is ready for reading from the network reqardless. So only improvement is when we are waiting for the data to be available ( or during network lag) there is no hanging feeling.

Sure. But you requested that big lump of data, so you need to process it anyway. And nobody forces you to process all of it at once, you can as well read it in smaller chunks and give control back to the event loop in between.

This approach basically just avoids your process getting stuck on a blocking system call (which is the core of your problem).

**Comment by wongck68** [comment]
> Sure. But you requested that big lump of data first place, so you probably need to process it anyway. And nobody forces you to process all of it at once, you can as well read it in smaller chunks and give control back to the event loop in between.
Yeap, agreed. Have to in long processing. Did that in my Google maps program, giving the user a visual of % done while downloading from Google and while decoding the graphics. So these post no issues as they can be easily taken care of.

And absolutely correct in the process stuck on blocking part of the network.
